File: neural_network/util/partitioner.py
import math
import random
from typing import List


class Partitioner:
    """Class to randomly partition `n` integers into sets of size `m`
    """

    def __init__(self, n: int, m: int):
        """Constructor method

        Parameters
        ----------
        n : int
            Number of integers
        m : int
            Size of each set
        """
        if n <= 0 or m <= 0:
            raise ValueError(f"n ({n}) and m ({m}) must be positive integers")
        if m > n:
            raise ValueError(f"m ({m}) cannot be greater than n ({n})")
        self._n = n
        self._m = m

    def __call__(self) -> List[List[int]]:
        """Shuffles all integers from 0 to `n` - 1 and creates a partition of
        this list.

        Returns
        -------
        List[List[int]]
            The partitioned list
        """
        ints = list(range(self._n))
        random.shuffle(ints)
        output_list = []
        num_sets = math.ceil(self._n / self._m)
        for i in range(num_sets - 1):
            output_list.append(ints[i * self._m:(i + 1) * self._m])
        output_list.append(ints[(num_sets - 1) * self._m:])
        return output_list

File: javaNeuralNetwork/src/main/java/neural_network/util/Partitioner.java
package neural_network.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.stream.IntStream;

/** Class to randomly partition {@code numInts} integers into sets of size
 * {@code setSize}.
 *
 */
public class Partitioner {

    private final int numInts;
    private final int setSize;
    private Random random = new Random();

    /** Constructor method.
     *
     * @param numInts Number of integers.
     * @param setSize Size of each set.
     */
    public Partitioner(int numInts, int setSize) {
        if (numInts <= 0 || setSize <= 0) {
            throw new IllegalArgumentException(
                    "numInts (%d) and setSize (%d) must be positive integers."
                            .formatted(numInts, setSize));
        } else if (setSize > numInts) {
            throw new IllegalArgumentException(
                    "setSize (%d) cannot be greater than numInts (%d)."
                            .formatted(setSize, numInts));
        }
        this.numInts = numInts;
        this.setSize = setSize;
    }

    /** Shuffles all integers from {@code 0} to {@code numInts - 1} and creates
     * a partition of this list.
     *
     * @return The partitioned list.
     */
    public List<List<Integer>> call() {
        List<Integer> ints = new ArrayList<>(IntStream.range(0, numInts)
                .boxed().toList());
        Collections.shuffle(ints, random);
        List<List<Integer>> outputList = new ArrayList<>();
        int numSets = (int) Math.ceil((double) numInts / setSize);
        for (int i = 0; i < numSets - 1; i ++) {
            outputList.add(ints.subList(i * setSize, (i + 1) * setSize));
        }
        outputList.add(ints.subList((numSets - 1) * setSize, numInts));
        return outputList;
    }

    public void setRandom(Random random) {
        this.random = random;
    }
}
